File: examples3d/debug_infinite_fall3.rs
use rapier_testbed3d::TestbedViewer;
use rapier3d::prelude::*;

pub async fn run(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    let ground_size = 100.1;
    let ground_height = 2.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(Vector::new(0.0, 4.0, 0.0));
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
    let (_handle, _) = world.insert(rigid_body, collider);

    let rad = 1.0;
    // Build the dynamic box rigid body.
    let rigid_body = RigidBodyBuilder::dynamic()
        .translation(Vector::new(0.0, 7.0 * rad, 0.0))
        .can_sleep(false);
    let collider = ColliderBuilder::ball(rad);
    let (_handle, _) = world.insert(rigid_body, collider);

    let rigid_body = RigidBodyBuilder::dynamic()
        .translation(Vector::new(0.0, 2.0 * rad, 0.0))
        .can_sleep(false);
    let collider = ColliderBuilder::ball(rad);
    let (_handle, _) = world.insert(rigid_body, collider);

    /*
     * Set up the testbed.
     */
    viewer.look_at(Vec3::new(100.0, -10.0, 100.0), Vec3::ZERO);
    viewer.set_world(&mut world);

    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            world.step();
        }
    }
    Ok(())
}

File: python/rapier-testbed/rapier_testbed/examples3/debug_infinite_fall3.py
"""Port of ``examples3d/debug_infinite_fall3.rs``.

A ground floating above the origin with two balls dropped from below it.
Reproduces a corner-case where bodies that start under the ground end up
falling indefinitely.
"""
from __future__ import annotations

import rapier3d as rp

from .._registry import register

CATEGORY = "Debug"
NAME = "Infinite fall"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 100.1
    ground_height = 2.1
    ground_body = rp.RigidBody.fixed().translation((0.0, 4.0, 0.0))
    ground_handle = bodies.insert(ground_body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size),
        ground_handle,
        bodies,
    )

    rad = 1.0
    for y in (7.0 * rad, 2.0 * rad):
        body = (
            rp.RigidBody.dynamic()
            .translation((0.0, y, 0.0))
            .can_sleep(False)
        )
        handle = bodies.insert(body)
        colliders.insert_with_parent(rp.Collider.ball(rad), handle, bodies)

    testbed.look_at((100.0, -10.0, 100.0), (0.0, 0.0, 0.0))
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
